extensions/libarchive/WriteArchiveStream.cpp

```diff
@@ -0,0 +1,151 @@
+    result = archive_write_add_filter_xz(arch.get());
+    if (result != ARCHIVE_OK) {
+      logger_->log_error("Archive write add filter xz error %s", archive_error_string(arch.get()));
+      return nullptr;
+    }
+  } else {
+    logger_->log_error("Archive write unsupported compression format");
+    return nullptr;
+  }
+  result = archive_write_set_bytes_per_block(arch.get(), 0);
+  if (result != ARCHIVE_OK) {
+    logger_->log_error("Archive write set bytes per block error %s", archive_error_string(arch.get()));
+    return nullptr;
+  }
+  result = archive_write_open(arch.get(), sink_.get(), nullptr, archive_write, nullptr);
+  if (result != ARCHIVE_OK) {
+    logger_->log_error("Archive write open error %s", archive_error_string(arch.get()));
+    return nullptr;
+  }
+  return arch;
+}
+
+bool WriteArchiveStreamImpl::newEntry(const EntryInfo& info) {
+  if (!arch_) {
+    return false;
+  }
+  arch_entry_ = archive_entry_new();
+  if (!arch_entry_) {
+    logger_->log_error("Failed to create archive entry");
+    return false;
+  }
+  archive_entry_set_pathname(arch_entry_.get(), info.filename.c_str());
+  archive_entry_set_size(arch_entry_.get(), info.size);
+  archive_entry_set_mode(arch_entry_.get(), S_IFREG | 0755);
+
+  int result = archive_write_header(arch_.get(), arch_entry_.get());
+  if (result != ARCHIVE_OK) {
+    logger_->log_error("Archive write header error %s", archive_error_string(arch_.get()));
+    return false;
+  }
+  return true;
+}
+
+size_t WriteArchiveStreamImpl::write(const uint8_t* data, size_t len) {
+  if (!arch_ || !arch_entry_) {
+    return STREAM_ERROR;
+  }
+
+  int result = archive_write_data(arch_.get(), data, len);
+  if (result < 0) {
+    logger_->log_error("Archive write data error %s", archive_error_string(arch_.get()));
+    arch_entry_.reset();
+    arch_.reset();
+    return STREAM_ERROR;
+  }
+
+  return len;
+}
+
+class ArchiveStreamProviderImpl : public ArchiveStreamProvider {
+ public:
+  using ArchiveStreamProvider::ArchiveStreamProvider;
+  std::unique_ptr<WriteArchiveStream> createWriteStream(int compress_level, const std::string& compress_format,
+                                              std::shared_ptr<OutputStream> sink, std::shared_ptr<core::logging::Logger> logger) override {
+    CompressionFormat format = CompressionFormat::parse(compress_format.c_str(), CompressionFormat{});
+    if (!format) {
+      if (logger) {
+        logger->log_error("Unrecognized compression format '%s'", compress_format);
+      }
+      return nullptr;
+    }
+    return std::make_unique<WriteArchiveStreamImpl>(compress_level, format, std::move(sink));
+  }
+
+  std::unique_ptr<ReadArchiveStream> createReadStream(std::shared_ptr<InputStream> archive_stream) override {
+    return std::make_unique<ReadArchiveStreamImpl>(std::move(archive_stream));
+  }
+};
+
+REGISTER_INTERNAL_RESOURCE_AS(ArchiveStreamProviderImpl, ("ArchiveStreamProvider"));
```

Comment:
We should move this to a separate file to avoid a dependency between WriteArchiveStream -> ReadArchiveStream. It's fine for me if it has an empty header, or just a forward declaration of the class, or no header at all.

Reply:
moved it to a separate source